### PR TITLE
Statsd Exporter testing

### PR DESCRIPTION
**Pull request by stefan-chivu**

### CODE REVIEW

File: gateway/exporters/statsd/statsd.go [hunk 1]
```diff
@@ -115,10 +115,11 @@ func (e *StatsdExporter) Export(leaf *ctree.Leaf) {
 					fieldVal, exists := targetConfig.Meta[fieldName]
 					if exists {
 						point.Tags[fieldName] = fieldVal
-					} else {
-						e.config.Log.Error().Msg("Field " + fieldName + " is not set for device " + point.Tags["target"] + " but is listed in the exporter's Metadata allow-list")
-						return
 					}
+					// else {
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd.go]: We should clean this up.

File: gateway/exporters/statsd/statsd_test.go [hunk 2]
```diff
@@ -1,57 +1,36 @@
...
 	pb "github.com/openconfig/gnmi/proto/gnmi"
-	"github.com/openconfig/gnmi/proto/target"
 	"github.com/stretchr/testify/assert"
 
 	"github.com/openconfig/gnmi-gateway/gateway/exporters"
 )
 
 var _ exporters.Exporter = new(StatsdExporter)
 var serverAddress = "127.0.0.1:8125"
+var zookeeperAddress = "172.17.0.2:2181"
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd_test.go]: we could use an env variable for the zk address

File: gateway/exporters/statsd/statsd_test.go [hunk 3]
```diff
@@ -64,86 +43,167 @@ func TestStatsdExporter_Name(t *testing.T) {
...
+			}
+
+		}
+
+	}()
+	assert.NotPanics(t, func() {
+		err = e.Start(&connMgr)
+		assert.Nil(t, err)
+
+		assert.Nil(t, err)
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd_test.go]: nit: duplicate

File: gateway/exporters/statsd/statsd_test.go [hunk 4]
```diff
@@ -64,86 +43,167 @@ func TestStatsdExporter_Name(t *testing.T) {
...
+		statsdServerConn, _ := createStatsdServer()
+		buf := make([]byte, 1024)
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				n, addr, _ := statsdServerConn.ReadFromUDP(buf)
+				t.Log("Received ", string(buf[0:n]), " from ", addr)
+			}
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd_test.go]: we'll need to add some assertions on the received metrics

File: gateway/exporters/statsd/statsd.go [hunk 5]
```diff
@@ -147,34 +145,22 @@ func (e *StatsdExporter) Export(leaf *ctree.Leaf) {
...
-		case float32:
-			if err := e.client.Gauge(string(metricJSON), int64(metric.Value.(float32)), 1); err != nil {
+		val, isNumericValue := utils.GetNumberValues(update.Val)
+		if isNumericValue {
+			if err := e.client.Gauge(string(metricJSON), int64(val), 1); err != nil {
 				e.config.Log.Error().Msg(err.Error())
 			}
-		case string:
-			if err := e.client.Set(string(metricJSON), string(metric.Value.(string)), 1); err != nil {
+		} else if valType := fmt.Sprintf("%T", metric.Value); valType == "*gnmi.TypedValue_StringVal" {
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd.go]: this is quite inefficient, can we check the type without using fmt?

File: gateway/exporters/statsd/statsd_test.go [hunk 6]
```diff
@@ -64,86 +52,179 @@ func TestStatsdExporter_Name(t *testing.T) {
...
+				n, _, _ := statsdServerConn.ReadFromUDP(buf)
+				output.s = append(output.s, string(buf[0:n]))
+			}
+		}
+	}()
+	assert.NotPanics(t, func() {
+		err = e.Start(&connMgr)
+		assert.Nil(t, err)
+
+		time.Sleep(50 * time.Millisecond)
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd_test.go]: relying solely on time.Sleep usually leads to flaky tests but it's probably ok for now.

File: gateway/exporters/statsd/statsd_test.go [hunk 7]
```diff
@@ -64,86 +52,179 @@ func TestStatsdExporter_Name(t *testing.T) {
...
 
-	zkConn, _, err := zk.Connect(config.ZookeeperHosts, config.ZookeeperTimeout)
+	done := make(chan bool, 1)
+	output := &MutexStringSlice{
+		s: []string{},
+	}
 
-	connMgr, err = connections.NewZookeeperConnectionManagerDefault(config, zkConn, nil)
+	go func() {
+		output.mu.Lock()
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd_test.go]: This mutex isn't very useful right now. There's a single writer that adds one item at a time, while the slice reader doesn't use it. You may want to use the mutex when accessing the slice.

File: gateway/exporters/statsd/statsd_test.go [hunk 8]
```diff
@@ -52,6 +52,7 @@ func TestStatsdExporter_Name(t *testing.T) {
 
 func TestStatsdExporter_Export(t *testing.T) {
 
+	os.Setenv("ZOOKEEPER_HOSTS", "172.17.0.2")
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd_test.go]: why would we set this env variable here? the point was to allow the one running the tests to provide the address, which may be different. If we're setting it here, we're overriding user values.

File: gateway/exporters/statsd/statsd.go [hunk 9]
```diff
@@ -155,7 +156,7 @@ func (e *StatsdExporter) Export(leaf *ctree.Leaf) {
 			if err := e.client.Gauge(string(metricJSON), int64(val), 1); err != nil {
 				e.config.Log.Error().Msg(err.Error())
 			}
-		} else if valType := fmt.Sprintf("%T", metric.Value); valType == "*gnmi.TypedValue_StringVal" {
+		} else if valType := reflect.TypeOf(metric.Value).String(); valType == "*gnmi.TypedValue_StringVal" {
```
Comment (petrutlucian94) [inline gateway/exporters/statsd/statsd.go]: nit: usually you do something like this as opposed to comparing strings:

```
reflect.TypeOf(metric.Value) == reflect.TypeOf((*gnmi.TypedValue_StringVal)(nil))
```

File: gateway/exporters/statsd/statsd_test.go (hunk 8 above)
Comment (stefan-chivu) [reply gateway/exporters/statsd/statsd_test.go]: I used it for my env and forgot to remove it. Fixed

**Pull request merged**